### ISSUE TITLE
new quanteda R

### ISSUE THREAD

**Issue by johnzfeng** [kbenoit/quanteda#934]
Hi, I have used quanteda and built a textual analysis tool in R, using previous version of R and quanteda. 

These are the codes:
#thisCaseSet is the data I use, which contains text information.
training <- thisCaseSet[thisNum$training, ]
testing <- thisCaseSet[thisNum$testing, ]

#get_q_corpus and get_q_dfm is the wrapper that call corpus and dfm of quanteda to generate corpus or dfm from my data (thisCaseSet).
 
q_corpus <- get_q_corpus (training, thisDataType)
q_dfm_training <- get_q_dfm (q_corpus, thisType = "tfidf")
q_corpus <- get_q_corpus (testing, thisDataType)
q_dfm_testing <- get_q_dfm (q_corpus, thisType = "tfidf")

#I use reason as label
thisModel <- textmodel(x = q_dfm_training, y = training$reason, model = "NB")

q_dfm_testing <- dfm_select(q_dfm_testing, q_dfm_training)
thisPredictionNB <- predict(thisModel, newdata = q_dfm_testing)

#end of code.

The codes worked fine. Recently, 8/21/2017, I have to update R and hence to new quanteda.  The codes now fail on  q_dfm_testing <- dfm_select(q_dfm_testing, q_dfm_training). The error message is:

Error in names(x) <- value : 
  'names' attribute [1] must be the same length as the vector [0]

I then change this line of code q_dfm_testing <- dfm_select(q_dfm_testing, q_dfm_training)
to 
q_dfm_testing <- dfm_select(q_dfm_testing, featnames(q_dfm_training))

This solved the error, but code
thisPredictionNB <- predict(thisModel, newdata = q_dfm_testing) now generate this error message:

Error in predict.textmodel_NB_fitted(thisModel, newdata = q_dfm_testing) : 
  feature set in newdata different from that in training set

I do not understand, and hope some one can help.  Thanks.

**Comment by kbenoit** [comment]
Hard to diagnose without being reproducible, so could you send us links to:
* `thisModel`
* `q_dfm_testing`
* `q_dfm_training`

**Comment by johnzfeng** [comment]
wow, that was a quick response.  Quanteda is really nice tool anyway. Here are the object you are asking, written with saveRDS function.

https://onedrive.live.com/?id=8D297E7485F2A755%21155&cid=8D297E7485F2A755

**Comment by johnzfeng** [comment]
I used the following three lines of code:

features <- Reduce(intersect, list(featnames(q_dfm_testing),featnames(q_dfm_training)))
q_dfm_testing <- dfm_select(q_dfm_testing, features)
q_dfm_training <- dfm_select(q_dfm_training, features)

to replace: q_dfm_testing <- dfm_select(q_dfm_testing, q_dfm_training).

This works. However, I have a question regarding using quanteda as a basis for production of my codes.  

1) I used q_dfm_training and its labels to obtain a model called thisModel
2) I saved thisModel at AWS S3
3) New data q_dfm_newdata comes in and this data has no label
4) q_dfm_training has features F1 while q_dfm_newdata has features F2.  
5) F1 and F2 have overlapped features but also have features distinct from each other.
6) I want to predict q_dfm_training's labeling based on F2 and thisModel, however, thisModel is fixed and so is F1.
7) Can I by pass this problem using Quanteda?  I know naïve bayes function in package "e1071" allows me to do that. I want to use Quanteda because it is nice.   
8) If Quanteda can also get it done, please advise how can I get it done? 

Thanks.

**Comment by koheiw** [comment]
Try `dfm_select(q_dfm_testing, pattern = q_dfm_training)` to get a dfm identical to the training data.

**Comment by johnzfeng** [comment]
Thanks, Koheiw.

I tried 

q_dfm_testing <- dfm_select(q_dfm_testing, pattern = q_dfm_training)

and the same error message as using:
q_dfm_testing <- dfm_select (q_dfm_testing, q_dfm_training)

Error in names(x) <- value : 
  'names' attribute [1] must be the same length as the vector [0]

I wonder it is a bug of dfm_select.

**Comment by koheiw** [comment]
OK. It is possible a bug. Can you share your q_dfm_testing and q_dfm_training with us, so that we can investigate?

**Comment by koheiw** [comment]
I created a branch that should solve the problem. Can you try **hotfix-issue934**?

**Comment by johnzfeng** [comment]
Dear Kbenoit,

Thanks for working hard into my issue.  

I reinstalled quanteda, by following https://github.com/kbenoit/readtext/blob/master/README.Rmd, and run the codes listed in this post again.  Codes finished as before issue arisen, i.e., there is no error message reported.

Quanteda is a great tool.  I am studying its new add-ons (topic model, etc), and look forward to more features.

**Comment by kbenoit** [comment]
Great to hear. If you like **quanteda** please consider leaving a testimonial at https://github.com/kbenoit/quanteda/issues/461
